compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateNoSetStateInRender.ts

```diff
@@ -155,20 +261,48 @@ function validateNoSetStateInRenderImpl(
                 }),
               );
             } else if (unconditionalBlocks.has(block.id)) {
-              errors.pushDiagnostic(
-                CompilerDiagnostic.create({
-                  category: ErrorCategory.RenderSetState,
-                  reason:
-                    'Calling setState during render may trigger an infinite loop',
-                  description:
-                    'Calling setState during render will trigger another render, and can lead to infinite loops. (https://react.dev/reference/react/useState)',
-                  suggestions: null,
-                }).withDetails({
-                  kind: 'error',
-                  loc: callee.loc,
-                  message: 'Found setState() in render',
-                }),
-              );
+              let isArgPrimitive = false;
+
+              if (instr.value.args.length > 0) {
+                const arg = instr.value.args[0];
+                if (arg.kind === 'Identifier') {
+                  isArgPrimitive = isPrimitiveSetArg(arg, fn);
```

Comment:
I expected that this would be something like: `if (!arg.reactive && HIR.isPrimitiveType(arg.identifier)) { /* allow */ }` - ie reusing type inference to determine if a value is a primitive. And likely combining with a `!arg.reactive` check because if the value being set is reactive, then it could still infinite loop even if its a primitive.